### PR TITLE
Additional built-in container styles

### PR DESCRIPTION
**Pull request by carlos-zamora**
Added `good`, `attention`, `warning`, and `accent` to container styles.
**Spec**: #1886 
**Components**:
* .NET (#2214 )
* Android (#2213)
* iOS (#2212)
* UWP (#2211)
* Shared (#2210)

# Testing
* Updated HostConfigs to include new Container Style Definitions
* Added AdaptiveCards of name `Container.Style.<stylename>.json`

# .NET
* `GetContainerStyleConfig()` used to help get desired style definition
Note: Bug #2285 is still present and makes ForegroundColors not work

# Shared Model
* Added to Enums and HostConfig

# UWP
* UWP Test App results
* Main changes made to `AdaptiveContainerStylesDefinition.h/cpp` and `Util.cpp`
* `GetContainerStyleDefinition()` used to help get desired style definition

# Android
* Expanded control flow to switch-case statement in ContainerRenderer and TextBlockRenderer
Note: FactSet just redirects to pertinent portion of TextBlockRenderer

# iOS
* Added `getColorPalette()` to get  desired style definition easier
* Fixed `sample.json`
* Fixed `getTextBlockColor()`

### CODE REVIEW

File: source/shared/cpp/ObjectModel/HostConfig.h [hunk 1]
```diff
@@ -185,6 +185,10 @@ namespace AdaptiveSharedNamespace
                                                         {"#FFFFD700", "#B2FFD700"}, // warning
                                                         {"#FF8B0000", "#B28B0000"}  // attention
                                                     }};
+        ContainerStyleDefinition goodPalette;
+        ContainerStyleDefinition attentionPalette;
+        ContainerStyleDefinition warningPalette;
+        ContainerStyleDefinition accentPalette;
```
Comment (RebeccaAnne) [inline source/shared/cpp/ObjectModel/HostConfig.h]: Do we want default values for these?

File: source/uwp/AdaptiveCardTestApp/Expected/microsoft-teams.Tests.Action.DuplicateIds.json [hunk 2]
```diff
@@ -1 +1 @@
-{"HostConfigHash":"5da68b8","CardHash":"e423716","Error":"Parsing card failed"}
\ No newline at end of file
+{"HostConfigHash":"1eae351","CardHash":"e423716","Error":"Parsing card failed"}
```
Comment (RebeccaAnne) [inline source/uwp/AdaptiveCardTestApp/Expected/microsoft-teams.Tests.Action.DuplicateIds.json]: Do you know why all these files changed?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm [hunk 3]
```diff
@@ -365,28 +365,54 @@ + (UIColor *)convertHexColorCodeToUIColor:(const std::string&)hexColorCode
     return color;
 }
 
-- (UIColor *)getBackgroundColorForContainerStyle:(ACRContainerStyle)style
+- (ContainerStyleDefinition) getColorPalette:(ACRContainerStyle)style
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: you are returning a struct by value. i suggest adding another param to diff the request either to foreground or background and return only those string instead of whole structure.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm [hunk 4]
```diff
@@ -365,28 +365,54 @@ + (UIColor *)convertHexColorCodeToUIColor:(const std::string&)hexColorCode
...
+- (UIColor *)getBackgroundColorForContainerStyle:(ACRContainerStyle)style
+{
+    std::string hexColorCode = [self getColorPalette:style].backgroundColor;
     return [ACOHostConfig convertHexColorCodeToUIColor:hexColorCode];
 }
 
 + (ACRContainerStyle)getPlatformContainerStyle:(ContainerStyle)style
 {
-    ACRContainerStyle containerStyle = ACRDefault;
+    ACRContainerStyle containerStyle;
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: why did you remove the initialization with the default value?

Comment (carlos-zamora) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: The default value is assigned in the `default` case for the switch-case. Felt it would be redundant to initialize it, but should I change it?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm (hunk 3 above)
Comment (carlos-zamora) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: The problem is that the background is a `string` whereas the foreground is a `ColorsConfig` struct. Having this helper method lets me get the background through `getBackgroundColorForContainerStyle()` (in an appropriately converted format) and the foreground colors directly.

File: source/shared/cpp/ObjectModel/HostConfig.h (hunk 1 above)
Comment (carlos-zamora) [reply source/shared/cpp/ObjectModel/HostConfig.h]: As of now, the default value is the same as the predefined `defaultPalette` (NOT the one defined by the user in the imported HostConfig). @matthidinger @andrewleader , thoughts on proper behavior or default values for this?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm (hunk 4 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: yes, assignment takes only 2-3 instructions.

---
In reply to: [249849267](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r249849267) [](ancestors = 249849267)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRTextBlockRenderer.mm [hunk 5]
```diff
@@ -81,11 +81,8 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
 
         // Obtain text color to apply to the attributed string
         ACRContainerStyle style = lab.style;
-        ColorsConfig colorConfig;
-        if (style == ACREmphasis)
-            colorConfig = config->GetContainerStyles().emphasisPalette.foregroundColors;
```
Comment (nesalang) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRTextBlockRenderer.mm]: need to change to have reference here. I didn't realize that you've changed this in earlier commit.

**Comment by nesalang** [comment]
>         ContainerStylesDefinition GetContainerStyles() const;

every usages of this method, user later choose containerstyledefintion based on style. it would be better add style param instead of returning whole container styles definition.

---
Refers to: source/shared/cpp/ObjectModel/HostConfig.h:282 in d8cbc59. [](commit_id = d8cbc59f300392900cb2d6b49ef5de398b40de27, deletion_comment = False)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm (hunk 3 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: can you make it return reference?

---
In reply to: [249862786](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r249862786) [](ancestors = 249862786)

File: source/shared/cpp/ObjectModel/HostConfig.h (hunk 1 above)
Comment (andrewleader) [reply source/shared/cpp/ObjectModel/HostConfig.h]: We have a default value for emphasisPalette, correct? I think we should similarly specify default values for goodPalette/etc

Comment (carlos-zamora) [reply source/shared/cpp/ObjectModel/HostConfig.h]: Sure. What do you want them to be? In the tests, I have them set to the subtle color for each ContainerStyle text.

**Comment by carlos-zamora** [comment]
Added `GetContainerStyle(style)`, `GetBackgroundColor(style)`, and `GetForegroundColor(style)` to move some of the repeated logic back to the SharedModel. Should look a lot cleaner now. Can't make use of this in UWP though because UWP has it's own weird ObjectModel on top of it.

File: source/shared/cpp/ObjectModel/HostConfig.cpp [hunk 6]
```diff
@@ -565,6 +577,36 @@ unsigned int HostConfig::GetFontWeight(FontStyle style, TextWeight weight) const
     return result;
 }
 
+ContainerStyleDefinition HostConfig::GetContainerStyle(ContainerStyle style) const
```
Comment (nesalang) [inline source/shared/cpp/ObjectModel/HostConfig.cpp]: Could you please make changes so that it wouldn't return a big struct?

File: source/shared/cpp/ObjectModel/HostConfig.cpp [hunk 7]
```diff
@@ -565,6 +577,36 @@ unsigned int HostConfig::GetFontWeight(FontStyle style, TextWeight weight) const
...
+            return _containerStyles.defaultPalette;
+    }
+}
+
+std::string HostConfig::GetBackgroundColor(ContainerStyle style) const
+{
+    return GetContainerStyle(style).backgroundColor;
+}
+
+ColorsConfig HostConfig::GetForegroundColors(ContainerStyle style) const
```
Comment (nesalang) [inline source/shared/cpp/ObjectModel/HostConfig.cpp]: Could you look into returnning ColorsConfig also?

File: source/shared/cpp/ObjectModel/HostConfig.cpp (hunk 6 above)
Comment (carlos-zamora) [reply source/shared/cpp/ObjectModel/HostConfig.cpp]: So, the problem with returning a pointer in the SharedModel is that all of the other similar functions in HostConfig (i.e.: `GetImageSizes()`, `GetImageBaseUrl()`, etc.) return the struct too. @paulcam206 was saying that passing by reference starts to cause issues with the lifetime of the pointer. I'm hoping `GetBackgroundColor(style)` and `GetForegroundColors(style)` help with that.  If not, suggestions with how to tackle this?

Comment (nesalang) [reply source/shared/cpp/ObjectModel/HostConfig.cpp]: one suggestion is leaving this function as it is, keep the copy of the returned value with ACOHostConfig, and move GetBackGroundColor and GetForegroundColor to platform codes. 
in case of iOS, GetBackgroundColor and GetForegroundColor could directly access the copied value. in that way we keep the copying as one time cost. 
And we can also remove ColorsConfig param from getTextBlockColor after making it an instance function.

---
In reply to: [250766675](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r250766675) [](ancestors = 250766675)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm (hunk 3 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOHostConfig.mm]: can get rid of this fuction by making getTextBlock color an instance method and access color config through its host config and add optional style param.

---
In reply to: [249968877](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r249968877) [](ancestors = 249968877,249862786)

**Comment by nesalang** [comment]
iOS looks good :)

File: source/uwp/AdaptiveCardTestApp/Expected/microsoft-teams.Tests.Action.DuplicateIds.json (hunk 2 above)
Comment (carlos-zamora) [reply source/uwp/AdaptiveCardTestApp/Expected/microsoft-teams.Tests.Action.DuplicateIds.json]: HostConfigs got updated to include more ContainerStyleDefinitions. 

---
In reply to: [247212915](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r247212915) [](ancestors = 247212915)

File: source/uwp/Renderer/lib/Util.cpp [hunk 8]
```diff
@@ -717,24 +717,47 @@ HRESULT GetColorFromString(const std::string& colorString, _Out_ ABI::Windows::U
 }
 CATCH_RETURN;
 
-HRESULT GetColorFromAdaptiveColor(_In_ ABI::AdaptiveNamespace::IAdaptiveHostConfig* hostConfig,
+HRESULT GetContainerStyleDefinition(ABI::AdaptiveNamespace::ContainerStyle style,
+                                    _In_ ABI::AdaptiveNamespace::IAdaptiveHostConfig* hostConfig,
+                                    _Out_ ABI::AdaptiveNamespace::IAdaptiveContainerStyleDefinition** styleDefinition) noexcept try
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/Util.cpp]: _Outptr_

File: source/uwp/Renderer/lib/Util.cpp [hunk 9]
```diff
@@ -717,24 +717,47 @@ HRESULT GetColorFromString(const std::string& colorString, _Out_ ABI::Windows::U
...
+    case ABI::AdaptiveNamespace::ContainerStyle::Default:
+    default:
+        RETURN_IF_FAILED(containerStyles->get_Default(styleDefinition));
+        break;
+    }
+    return S_OK;
+}
+CATCH_RETURN;
+
+HRESULT GetColorFromAdaptiveColor(ABI::AdaptiveNamespace::IAdaptiveHostConfig* hostConfig,
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/Util.cpp]: _In_

File: source/dotnet/Library/AdaptiveCards/AdaptiveContainerStyle.cs [hunk 10]
```diff
@@ -20,6 +20,14 @@ public enum AdaptiveContainerStyle
...
-        Emphasis = 1
+        Emphasis = 1,
+
+        Good = 2,
+
+        Attention = 3,
+
+        Warning = 4,
+
+        Accent = 5
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards/AdaptiveContainerStyle.cs]: nitpick -- I don't think we need the extra newlines here

File: source/ios/AdaptiveCards/ADCIOSVisualizer/resources/sample.json [hunk 11]
```diff
@@ -55,57 +55,160 @@
...
+					"default": "#2E89FC",
+					"subtle": "#882E89FC"
+				},
+				"good": {
+					"default": "#54a254",
+					"subtle": "#DD54a254"
+				},
+				"warning": {
+					"default": "#c3ab23",
+					"subtle": "#DDc3ab23"
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/resources/sample.json]: out of curiosity, why are we mixing case here?

Comment (carlos-zamora) [reply source/ios/AdaptiveCards/ADCIOSVisualizer/resources/sample.json]: Didn't realize it's not case sensitive. Oops!

---
In reply to: [257394613](https://github.com/Microsoft/AdaptiveCards/pull/2289#discussion_r257394613) [](ancestors = 257394613)
